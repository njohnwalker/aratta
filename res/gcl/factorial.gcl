@req( n >= 0 )
res, n0 := 1, 1 ;
do @inv()
[] n0 < n ->
    n0 := n0 + 1;
    res := res * n0;
od
@ens()
