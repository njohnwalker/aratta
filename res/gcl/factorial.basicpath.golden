Path 1:
  { PRE: n >= 0 }
  ; Substitute
      @res -> 1
      @n0 -> 1
  { POST: INVARIANT == PLACEHOLDER }
Path 2:
  { PRE: INVARIANT == PLACEHOLDER }
  ; Assume n0 < n
  ; Substitute
      @n0 -> n0 + 1
  ; Substitute
      @res -> res * n0
  { POST: INVARIANT == PLACEHOLDER }
Path 3:
  { PRE: INVARIANT == PLACEHOLDER }
  ; Assume ~ n0 < n
  { POST: true }